### PR TITLE
👷 Create installation package for windows / linux / macos

### PR DESCRIPTION
**Pull request by coliss86**
Use [jpackage](https://docs.oracle.com/en/java/javase/14/jpackage/basic-packaging.html#GUID-24F95A39-049C-491A-B49A-75898FAA7C57) to create native installer for each platform.
The binary produced is the jar of Makelangelo with a JVM embedded, no more painful installation of java 🎉

On mac os, it produced an DMG which looks like this: 
image

When launching the app, the icon in popup is not the luke but the icons of Makelangelo: 
image

Also, the icon in the dock and in the Applications folder are replaced:
image

The script `src/main/package/start OSX.command` will not be useful anymore. 
The same goes for linux and windows, I did not test them now.

TODO:

- [x] test on mac
- [x] test on ubuntu
- [x] test on windows
- [x] modify README.md

### CODE REVIEW

**Comment by PPAC37** [comment]
Superb idea.
This will avoid many complications for beginners.

I look forward to the progress. 
(I know this hides a lot of complications/difficulty to get there for each platform (linux and all its distributions are a headache) ... )

**Comment by coliss86** [comment]
Before making it ready for merge, can you @i-make-robots @PPAC37 try the artifact uploaded please ?
You'll find them here: https://github.com/MarginallyClever/Makelangelo-software/actions/runs/1831799824

**Comment by PPAC37** [comment]
Already, well done for not having given up, in the battle to find the right sequence of commands to achieve this (so many commits).

The .deb installed without error (on my distribution (Ubuntu 18.04) )
~~~
$ sudo dpkg -i makelangelo_7.31.0-1_amd64.deb 
Sélection du paquet makelangelo précédemment désélectionné.
(Lecture de la base de données... 571752 fichiers et répertoires déjà installés.)
Préparation du dépaquetage de makelangelo_7.31.0-1_amd64.deb ...
Dépaquetage de makelangelo (7.31.0-1) ...
Paramétrage de makelangelo (7.31.0-1) ...
~~~
(but for a program that does not need superuser rights, as it is a .deb to use `dpkg` or `gdebi` I need to do a `sudo`. (If there is the possibility of making an .AppImage and not a .deb we could avoid that ... ))

After that I can do (-nolf because I have a default Look And Feel system that bugs ...)
~~~
/opt/makelangelo/bin/Makelangelo -nolf
~~~
(I should be able to easily add it to my PATH environment variable, but not done, for the moment)
Which gets me started, Makelangelo. (however, no nice splashscreen when loading)

If I delete/rename my preference files
`mv ~/.java/.userPrefs ~/.java/.userPrefs12022022`
It does ask me at launch to choose a language (and I don't really know why, (order of an ls?) it offers me French as the first choice, also note that this is a dialog that does not create a icon (like the applications) in my system bar. But after validating and loading the main frame, I find an icon for the current application)
image

The few user tests that I have done seem to give the same results as with the .jar via my JVM (nothing to report, still the same small problems that I usually have and no new problem ... ex: connection which does not pass unless I once launch a connection with "Repetier host", ...)
image

So well done! for a .deb with an embedded JRE it's nickel.
(After for a beginner, it lacks a launcher integration into the system, but there it's always very tricky to find something that works, I'm used to it, like with .AppImage, to do it myself ...)

**Comment by coliss86** [comment]
Very nice feedback, it is not possible to create an .AppImage with the tool I'm using.

I have modified the creation to include the `-nolf`, creation of a shortcut menu in `Engineering` and adding a license, can you test again please ?
Binary: https://github.com/MarginallyClever/Makelangelo-software/actions/runs/1833344099

I still have to work on adding it in the PATH, it will be very useful

**Comment by PPAC37** [comment]
https://github.com/MarginallyClever/Makelangelo-software/actions/runs/1833344099
For me, perfect integration in the menu and has you have add the -nolf i do not have to change the .desktop file.

It ends up in "Autre"/"Other" because I'm with a custom menu that hides "Engineering".
image

**Comment by coliss86** [comment]
waou ! nice feedback :-) 
On your windows, can you please find the log file in order to fix the path

**Comment by coliss86** [comment]
thanks, I have created another PR https://github.com/MarginallyClever/Makelangelo-software/pull/540 to try to solve the admin issue.

**Comment by PPAC37** [comment]
I was wondering if it was possible to have a build number on the master branch. (and/or the commit hashes) embedded in the .jar ... because it seems to me that it can't be found ... but I haven't looked much either ...

**Comment by i-make-robots** [comment]
- [ ] "run app right now" button
- [ ] "add shortcut to desktop" checkbox
- [ ] any provisions for installing over an older copy?

**Comment by coliss86** [comment]
> * [ ]  "run app right now" button

can't do it

>     * [ ]  "add shortcut to desktop" checkbox

can't do it, here is the full doc of shortcuts: https://docs.oracle.com/en/java/javase/14/jpackage/manage-installation.html#GUID-832F848E-E074-476C-A203-F081D0407081

>     * [ ]  any provisions for installing over an older copy?

For windows, this is a "standard" msi installer, maybe it will ask you to overwrite the previous version.

**Comment by coliss86** [comment]
> > * [ ]  "add shortcut to desktop" checkbox
> 
> can't do it, here is the full doc of shortcuts: https://docs.oracle.com/en/java/javase/14/jpackage/manage-installation.html#GUID-832F848E-E074-476C-A203-F081D0407081

There an issue of this : https://bugs.openjdk.java.net/browse/JDK-8261536 => in a near future, it should be possible

**Comment by i-make-robots** [comment]
this is the correct logo.
icon_logo

**Comment by coliss86** [comment]
I have updated the logo

**Comment by coliss86** [comment]
Found an issue, converting it in draft to prevent merging

**Comment by coliss86** [comment]
Now it's ready

**Comment by i-make-robots** [comment]
still draft.

**Comment by coliss86** [comment]
now it is ready :)
Have a look at the README.md to fix any typo I made

File: README.md [hunk 1]
```diff
@@ -4,30 +4,36 @@ Makelangelo Software is a Java program that prepares art for CNC plotters.  It i
...
 
 # Start!
 
-- Setup [Marlin 3D printer firmware](https://github.com/MarginallyClever/Marlin-polargraph/commits/polargraph-2.0.x) with the help of the [documentation](https://www.marginallyclever.com/2021/10/friday-facts-4-how-to-marlin-polargraph/)
-- Install "Makelangelo Software" on your computer to control the robot 
-  1. Install The latest OpenJDK (java software)
-  2. Grab a [nightly build](https://github.com/MarginallyClever/Makelangelo-software/releases/tag/Nightly) and run it directly *or* Install Makelangelo Software App from the source
+- Buy a plotter [Makelangelo 5](https://www.marginallyclever.com/products/makelangelo-5/) or build your own;
+  ![plotter](src/main/resources/splash.png)
+- Setup [Marlin 3D printer firmware](https://github.com/MarginallyClever/Marlin-polargraph/commits/polargraph-2.0.x) with the help of the [documentation](https://www.marginallyclever.com/2021/10/friday-facts-4-how-to-marlin-polargraph/);
```
Comment (PPAC37) [inline README.md:21]: Maybe the tree https://github.com/MarginallyClever/Marlin-polargraph/tree/polargraph-2.0.x or directly the .zip https://github.com/MarginallyClever/Marlin-polargraph/archive/refs/heads/polargraph-2.0.x.zip
and not the commit page ?

File: src/main/package/jpackage-ubuntu-latest.cfg [hunk 2]
```diff
@@ -0,0 +1,5 @@
+--type deb
+--arguments "-nolf"
+--icon src/main/package/logo.ico
```
Comment (PPAC37) [inline src/main/package/jpackage-ubuntu-latest.cfg:3]: maybe a .png would be better for linux...

File: README.md (hunk 1 above)
Comment (coliss86) [reply README.md:21]: thanks, corrected

File: src/main/package/jpackage-ubuntu-latest.cfg (hunk 2 above)
Comment (coliss86) [reply src/main/package/jpackage-ubuntu-latest.cfg]: good, I changed that and added a png icon

**Comment by PPAC37** [comment]
Perfect for logo.png
In my menu
image

while running in my bar
image

But strangely not when I do an Alt+Tab... (but not really important)
image


For https://github.com/MarginallyClever/Marlin-polargraph/commits/polargraph-2.0.x (the commit to tree link,) the README.md have another one.

**Comment by coliss86** [comment]
> For https://github.com/MarginallyClever/Marlin-polargraph/commits/polargraph-2.0.x (the commit to tree link,) the README.md have another one.

Thanks, I forgot the second one, fixed now

**Comment by PPAC37** [comment]
In Makelangelo.jave Something like that, to have the ImageIcon when Alt+Tab 


~~~

private void createAppWindow() {
...

		try{
			// Allow to set a JFrame IconImage(s) (replace/overide the on defined by the lancher...)
			
			// // Test01 OK but transparence ... 
			//mainFrame.setIconImage(ImageIO.read(Makelangelo.class.getResource("/logo.png")));
			
			// // Test02 KO .ico give nothing ( need to by converted)			
			//mainFrame.setIconImage(ImageIO.read(Makelangelo.class.getResource("/favicon.ico")));			
			// // ?OK (not tested as it need a depandancy)  or to have an .ico decodeur like net.sf.image4j.codec.ico.ICODecoder to get a List<BufferedImage> ( a  List as an .ico file can containe multiple image with diffrent size  ... )
            //mainFrame.setIconImages(readIconesListFile(Makelangelo.class.getResource("/favicon.ico")));
			
			// // Test03 OK after under linux with imagemagik a `convert favicon.ico favicon.png` in src/main/resources
			//mainFrame.setIconImage(ImageIO.read(Makelangelo.class.getResource("/favicon.png")));
			
			// // Test04 OK logo.png from PR #536 renamed as MainFrameImageIcon.png and added in src/main/resources
			//
			mainFrame.setIconImage(ImageIO.read(Makelangelo.class.getResource("/MainFrameImageIcon.png")));
		}catch(Exception z){
			logger.debug("mainFrame setIconImage err : {}",z.getMessage(),z);
		}
...
}

// //Only if we whant to read .ico file.	
//	/**
//	 * Require : net.​sf.​image4j.​codec.​ico
//	 * <code>import net.sf.image4j.codec.ico.ICODecoder;</code>
//	 * @param fSub
//	 * @return 
//	 */
//    public static List<BufferedImage> readIconesListFile(File fSub) {
//        try {
//            List<BufferedImage> images = ICODecoder.read(fSub);
//            return images;
//        } catch (IOException e) {
//            System.err.printf("! reading %s : %s\n", fSub.getPath(), e.getMessage());
//        }
//        return null;
//    }

~~~

Test01
image

Test01 Alt Tab 
image

Test02
see previous message
Test02 Alt Tab 
see previous message

Test03
image

Test03 Alt Tab 
image


Test04
image

Test04 Alt Tab 
image

**Comment by coliss86** [comment]
Nice, I implemented test 4, can you give a try ?

**Comment by PPAC37** [comment]
> can you give a try ?
Perfect for me 
image

**Pull request merged**